File: python/017_letter_combinations.py
class Solution(object):
    def letterCombinations(self, digits):
        D = {'2': ['a', 'b', 'c'],
             '3': ['d', 'e', 'f'],
             '4': ['g', 'h', 'i'],
             '5': ['j', 'k', 'l'],
             '6': ['m', 'n', 'o'],
             '7': ['p', 'q', 'r', 's'],
             '8': ['t', 'u', 'v'],
             '9': ['w', 'x', 'y', 'z']}

        n = len(digits)

        if n == 0:
            return []

        strList = D[digits[0]].copy()
        total = len(strList)

        for i in range(1, n):
            m = len(strList)
            total *= len(D[digits[i]])
            for j in range(m):
                temp = strList[j]
                for k in D[digits[i]]:
                    strList.append(temp + k)

        return strList[-total:]


if __name__ == "__main__":
    s = Solution()
    print(s.letterCombinations("23"))

File: cpp/017_letter_combinations.cpp
#include "header.h"
using namespace std;

// backtracking 是 DFS 的一种表现形式
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) return {};
        vector< vector<char> > m_;
        m_ = {{' '}, {}, {'a', 'b', 'c'}, {'d', 'e', 'f'},\
        {'g', 'h', 'i'}, {'j', 'k', 'l'}, {'m', 'n', 'o'},\
        {'p', 'q', 'r', 's'}, {'t', 'u', 'v'}, {'w', 'x', 'y', 'z'}};
        string cur;
        vector<string> ans;
        dfs(digits, ans, cur, m_);
        return ans;
    }
private:
    void dfs(string& digits, vector<string>& ans, string& cur, vector< vector<char> > m_) {
        if (cur.size() == digits.size()) {
            ans.push_back(cur);
            return;
            }
        int num = digits[cur.size()]-'0';  // 小心记得转int
        for (char s : m_[num]) {
            cur += s;
            dfs(digits, ans, cur, m_);
            cur.pop_back();
        } 
    }
};

int main() {
    Solution s;
    string digits = "23";
    cout << s.letterCombinations(digits)[0] << endl;
}